### PR TITLE
UCP/UCT/API: Add API for non-blocking memory registration.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2775/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/195/ for details.

**Comment by yosefe** [comment]
@shamisp pls take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2164/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -171,6 +171,17 @@ enum ucp_dt_type {
 
 
 /**
+ * @ingroup UCP_MEM
+ * @brief UCP memory mapping flags.
+ *
+ * The enumeration list describes the datatypes supported by UCP.
```
Comment (shssf) [inline src/ucp/api/ucp.h]: It isn't clear why `datatypes` mentioned here.

Comment (shamisp) [inline src/ucp/api/ucp.h]: Agree - we need better documentation

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2788/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/211/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2185/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -174,10 +174,13 @@ enum ucp_dt_type {
  * @ingroup UCP_MEM
  * @brief UCP memory mapping flags.
  *
- * The enumeration list describes the datatypes supported by UCP.
+ * The enumeration list describes the memory mapping flags supported by @ref
+ * ucp_mem_map() function.
  */
 enum {
-    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0) /* Avoid populating the pages in the mapping up-front */
+    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0) /*< Complete the mapping faster, possible by
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: I would suggest elaborate even more. The actual mapping can be delayed to the communication request (or something like this).

**Comment by yosefe** [comment]
@shamisp fixed

**Comment by shamisp** [comment]
As far as I can see, we don't break ABI or API ? let's add special labels for API and ABI changes.

**Comment by yosefe** [comment]
it does not break because we already have flags parameter (unused so far)

**Comment by shamisp** [comment]
good

**Comment by shamisp** [comment]
So, this flag is essentially hint. A.k.a developers don't have to implement this.

**Comment by shamisp** [comment]
My point - if this is a hint, I would explicitly state it.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2789/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2790/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/212/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2186/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp pls take a look

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -178,9 +178,10 @@ enum ucp_dt_type {
  * ucp_mem_map() function.
  */
 enum {
-    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0) /*< Complete the mapping faster, possible by
-                                          avoiding population the pages in the mapping
-                                          up-front */
+    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0) /*< Complete the mapping faster, possibly by
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: /**<

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -178,9 +178,10 @@ enum ucp_dt_type {
...
+                                          not populating the pages in the mapping
+                                          up-front, and mapping them later when
+                                          they are accessed by communications. */
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: s/communication/communication routines/

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -340,9 +340,10 @@ enum {
  * @brief  Memory domain capability flags.
  */
 enum {
-    UCT_MD_MEM_FLAG_NONBLOCK = UCS_BIT(0),  /**< Non-blocking allocation/registration:
-                                                 Page mapping may be deferred until
-                                                 accessed by the CPU or a transport. */
+    UCT_MD_MEM_FLAG_NONBLOCK = UCS_BIT(0), /**< Hint to perform non-blocking
+                                                allocation/registration: page
+                                                mapping may be deferred until
+                                                accessed by the CPU or a transport. */
```
Comment (shamisp) [inline src/uct/api/uct.h]: it is accessed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/213/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2791/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2187/ for details (Mellanox internal link).

**Pull request merged**